// @poppinss/cliui/build/helpers-BVH1pXr1.js
import stringWidth from "string-width";
import cliTruncate from "cli-truncate";
import terminalSize from "terminal-size";
function createWordWrapper(start, stop) {
	const mode = "soft";
	const re = mode === "hard" ? /\b/ : /(\S+\s+)/;
	return function(text) {
		return text.toString().split(re).reduce(function(acc, x) {
			if (mode === "hard") for (let i = 0; i < x.length; i += stop - start) acc.push(x.slice(i, i + stop - start));
			else acc.push(x);
			return acc;
		}, []).reduce(function(lines, rawChunk) {
			if (rawChunk === "") return lines;
			const chunk = rawChunk.replace(/\t/g, "    ");
			const i = lines.length - 1;
			if (lines[i].length + chunk.length > stop) {
				lines[i] = lines[i].replace(/\s+$/, "");
				chunk.split(/\n/).forEach(function(c) {
					lines.push(new Array(start + 1).join(" ") + c.replace(/^\s+/, ""));
				});
			} else if (chunk.match(/\n/)) {
				const xs = chunk.split(/\n/);
				lines[i] += xs.shift();
				xs.forEach(function(c) {
					lines.push(new Array(start + 1).join(" ") + c.replace(/^\s+/, ""));
				});
			} else lines[i] += chunk;
			return lines;
		}, [new Array(start + 1).join(" ")]).join("\n");
	};
}
const TERMINAL_SIZE = terminalSize().columns;
function applyPadding(value, options) {
	if (options.paddingLeft) value = `${options.paddingChar.repeat(options.paddingLeft)}${value}`;
	if (options.paddingRight) value = `${value}${options.paddingChar.repeat(options.paddingRight)}`;
	return value;
}
function justify(columns, options) {
	const normalizedOptions = {
		align: "left",
		paddingChar: " ",
		...options
	};
	return columns.map((column) => {
		const columnWidth = stringWidth(column);
		if (columnWidth >= normalizedOptions.maxWidth) return column;
		if (normalizedOptions.align === "left") return applyPadding(column, {
			paddingChar: normalizedOptions.paddingChar,
			paddingRight: normalizedOptions.maxWidth - columnWidth
		});
		return applyPadding(column, {
			paddingChar: normalizedOptions.paddingChar,
			paddingLeft: normalizedOptions.maxWidth - columnWidth
		});
	});
}
function wrap(columns, options) {
	const wrapper = createWordWrapper(options.startColumn, options.endColumn);
	if (options.trimStart) return columns.map((column) => wrapper(column).trimStart());
	return columns.map((column) => wrapper(column));
}
function truncate(columns, options) {
	return columns.map((column) => cliTruncate(column, options.maxWidth, {
		truncationCharacter: options.truncationChar || "…",
		position: options.position || "end"
	}));
}
export { wrap as i, justify as n, truncate as r, TERMINAL_SIZE as t };
